classes/acceptances_table.php

```diff
@@ -393,13 +407,71 @@ protected function build_sql_for_roles_filter() {
         }
     }
 
+    /**
+     * Hook that can be overridden in child classes to wrap a table in a form
+     * for example. Called only when there is data to display and not
+     * downloading.
+     */
+    public function wrap_html_start() {
+        echo \html_writer::start_tag('form',
+            ['action' => new \moodle_url('/admin/tool/policy/accept.php'), 'data-action' => 'acceptmodal']);
+        echo \html_writer::empty_tag('input', ['type' => 'hidden', 'name' => 'sesskey', 'value' => sesskey()]);
+        echo \html_writer::empty_tag('input', ['type' => 'hidden', 'name' => 'returnurl',
+            'value' => $this->get_return_url()]);
+        foreach (array_keys($this->versionids) as $versionid) {
+            echo \html_writer::empty_tag('input', ['type' => 'hidden', 'name' => "versionids[{$versionid}]",
+                'value' => $versionid]);
+        }
+    }
+
+    /**
+     * Hook that can be overridden in child classes to wrap a table in a form
+     * for example. Called only when there is data to display and not
+     * downloading.
+     */
+    public function wrap_html_finish() {
+        global $PAGE;
+        if ($this->canagreeany) {
+            echo \html_writer::empty_tag('input', ['type' => 'submit',
+                'value' => get_string('agreetoselected', 'tool_policy'), 'class' => 'btn btn-primary']);
+            $PAGE->requires->js_call_amd('tool_policy/acceptmodal', 'getInstance', [\context_system::instance()->id]);
+        }
+        echo "</form>\n";
+    }
+
     /**
      * Render the table.
      */
     public function display() {
         $this->out(100, true);
     }
 
+    /**
+     * Call appropriate methods on this table class to perform any processing on values before displaying in table.
+     * Takes raw data from the database and process it into human readable format, perhaps also adding html linking when
+     * displaying table as html, adding a div wrap, etc.
+     *
+     * See for example col_fullname below which will be called for a column whose name is 'fullname'.
+     *
+     * @param array|object $row row of data from db used to make one row of the table.
+     * @return array one row for the table, added using add_data_keyed method.
+     */
+    public function format_row($row) {
+        $row->canaccept = false;
+        $row->select = null;
+        if (!$this->is_downloading()) {
+            \context_helper::preload_from_record($row);
+            if (has_capability('tool/policy:acceptbehalf', \context_system::instance()) ||
+                has_capability('tool/policy:acceptbehalf', \context_user::instance($row->id))) {
+                $row->canaccept = true;
+                $row->select = \html_writer::empty_tag('input',
```

Comment:
Should these checkboxes have some label for accessibility reasons? Currently they seem to be just visually bound to the row with the associated user.

Reply:
good point, again, I use the same syntax the new Participants page uses

Reply:
Thanks for adding it. Just to mention, I personally found the nested syntax `<label><input></label>` easier, especially in cases where there is no natural ID attribute of the input field. In this nested syntax, there is no need for the `for=` attribute.

Reply:
when label has class="accesshide" won't it hide the checkbox too in this case?

Reply:
Good point! The accesshide would have to be applied to a span holding the actual text inside the label, as in
```
<label>
    <span class="accesshide">Agree on behalf of John Smith</span>
    <input type="checkbox" />
</label>
```